### DISCUSSION TITLE
Curl Multi Handle:  issue, unable to provide report yet

### DISCUSSION THREAD

**Discussion by calvin2021y** [curl/curl#14252] [General]
I has a app use curl multi handle with libuv, after upgrade master branch (to prepare for  8.9 release), get some issue that not exists before. 

case 1:

fd pass to handle_socket with event (CURL_POLL_IN, CURL_POLL_OUT, CURL_POLL_INOUT),  socketp is NULL.  but the fd is assigned. (errors: file already exists) 

with event CURL_POLL_REMOVE,   socketp is NULL. 

The app had finish billions of requests for 3 years, I am not change the code unless upgrade curl.  

The problem is random and hard to create a test case.

This is the logs before crash:

```sh
* readwrite_data: we are done
* nread == 0, stream closed, bailing
* STATE: PERFORMING => DONE handle 0x5555564123c8; line 2549
* multi_done[DONE]: status: 0 prem: 0 done: 0
* Connection still in use 1, no more multi_done now!
* Expire cleared
* STATE: INIT => SETUP handle 0x5555563d1f18; line 1950
* STATE: SETUP => CONNECT handle 0x5555563d1f18; line 1966
* Found bundle for host: 0x555556380398 [can multiplex]
* Added connection 4. The cache now contains 2 members
* STATE: CONNECT => RESOLVING handle 0x5555563d1f18; line 1992
* HTTP/2 found, allow multiplexing
< HTTP/2 200 
< date: Mon, 22 Jul 2024 12:20:29 GMT
< content-type: application/json;charset=utf-8
< vary: Accept-Encoding
< content-encoding: gzip
< 
* readwrite_data: we are done
* nread == 0, stream closed, bailing
* STATE: PERFORMING => DONE handle 0x5555563ceab8; line 2549
* multi_done[DONE]: status: 0 prem: 0 done: 0
* multi_done, not reusing connection=1, forbid=0, close=1, premature=0, conn_multiplex=1
* The cache now contains 1 members
* Curl_disconnect(conn #1, aborted=0)
* shutting down connection #1
* shutdown start on connection
* [CCACHE] shutdown #1, done=0
* [CCACHE] set socket=15 events=1 on #1
```

I guess this related to @icing recent change.   (I am grateful for his great work made curl better).  I think maybe add libuv multi event test case will prevent this problem in future.

**Comment by icing** [comment]
Indeed, I think the shutdown implementation is incomplete in regards to event based processing. I will make a PR for this.

And you are right that the test coverage of this in more or less non existing. Some hero is needed.

**Comment by icing** [comment]
Could you check if #14257 improves your situation? Thanks!

**Comment by calvin2021y** [comment]
On it.

**Comment by calvin2021y** [comment]
after rebuilding and run 30 min:

```c
* readwrite_data: we are done
* nread == 0, stream closed, bailing
* STATE: PERFORMING => DONE handle 0x55555640b6e8; line 2549
* multi_done[DONE]: status: 0 prem: 0 done: 0
* multi_done, not reusing connection=0, forbid=0, close=1, premature=0, conn_multiplex=1
* The cache now contains 0 members
* Curl_disconnect(conn #0, aborted=0)
* shutting down connection #0
* shutdown start on connection
* [CCACHE] shutdown #0, done=0
* [CCACHE] added #0 to shutdown list of length 1
* Expire cleared
curl/lib/multi.c:3235: CURLMcode multi_socket(struct Curl_multi *, _Bool, curl_socket_t, int, int *): Assertion `data->magic == 0xc0dedbadU' failed.
```

**Comment by icing** [comment]
Thanks! I think I understood this problem and made an addition to the PR. Could you verify this change? Thanks!

**Reply by calvin2021y** [reply]
after 1 hours test,  seem stable.

I will let it run until tororrow and report again.

**Reply by calvin2021y** [reply]
after 11 hours, 5M successful request.  I guess the problem fixed.

**Reply by icing** [reply]
Thanks for testing this!

**Comment by bagder** [comment]
Fixed in git commit ae620a70a020f253c0d
